modules/javafx.graphics/src/main/java/com/sun/marlin/OffHeapArray.java

```diff
@@ -25,83 +25,134 @@
-    int  used;
+    private MemorySegment segment;
+    private long length;
+    private int used;
 
     OffHeapArray(final Object parent, final long len) {
+        arena = Arena.ofConfined();
+
         // note: may throw OOME:
-        this.address = UNSAFE.allocateMemory(len);
+        this.segment = arena.allocate(len, ALIGNMENT);
         this.length  = len;
         this.used    = 0;
-        if (LOG_UNSAFE_MALLOC) {
+        if (LOG_OFF_HEAP_MALLOC) {
             MarlinUtils.logInfo(System.currentTimeMillis()
                                 + ": OffHeapArray.allocateMemory =   "
-                                + len + " to addr = " + this.address);
+                                + len + " for segment = " + this.segment);
         }
 
         // Register a cleaning function to ensure freeing off-heap memory:
         MarlinUtils.getCleaner().register(parent, this::free);
     }
 
+    /**
+     * Gets the length of this array.
+     * @return the length in bytes
+     */
+    long getLength() {
+        return length;
+    }
+
+    /**
+     * Gets the number of bytes currently being used. Always <= length
+     * @return number of used bytes
+     */
+    int getUsed() {
+        return used;
+    }
+
+    /**
+     * Sets the number of bytes currently being used. Always <= length
+     * @param used number of used bytes
+     */
+    void setUsed(int used) {
+        this.used = used;
+    }
+
+    /**
+     * Increments the number of bytes currently being used.
+     * Current used + increment must be <= length
+     * @param increment number of used bytes to increment
+     */
+    void incrementUsed(int increment) {
+        this.used += increment;
+    }
+
     /*
      * As realloc may change the address, updating address is MANDATORY
      * @param len new array length
      * @throws OutOfMemoryError if the allocation is refused by the system
      */
     void resize(final long len) {
+        Arena newArena = Arena.ofConfined();
+
         // note: may throw OOME:
-        this.address = UNSAFE.reallocateMemory(address, len);
+        MemorySegment newSegment = newArena.allocate(len, ALIGNMENT);
+
+        // If there are any bytes in use, copy them to the newly reallocated array
+        if (this.used > 0) {
+            MemorySegment.copy(segment, 0, newSegment, 0, Math.min(this.used, len));
+        }
+
+        this.arena.close();
+        this.arena = newArena;
+        this.segment = newSegment;
         this.length  = len;
-        if (LOG_UNSAFE_MALLOC) {
+
+        if (LOG_OFF_HEAP_MALLOC) {
             MarlinUtils.logInfo(System.currentTimeMillis()
                                 + ": OffHeapArray.reallocateMemory = "
-                                + len + " to addr = " + this.address);
+                                + len + " for segment = " + this.segment);
         }
     }
 
     void free() {
-        UNSAFE.freeMemory(this.address);
-        if (LOG_UNSAFE_MALLOC) {
+        arena.close();
```

Comment:
since we don't assign arena to `null`, I assume that the code expects:

1. `free()` will be only called once
2. no access to this OffHeapArray happens after `free()` is called

right?